src/charm.py

```diff
@@ -165,12 +174,88 @@ def external_url(self) -> str:
     # === UTILITY METHODS === #
     ###########################
 
-    # FIXME: Move the alert_rules handling to Coordinator
-    def _consolidate_nginx_rules(self):
-        os.makedirs(CONSOLIDATED_ALERT_RULES_PATH, exist_ok=True)
-        os.makedirs(CONSOLIDATED_ALERT_RULES_PATH, exist_ok=True)
-        for filename in glob.glob(os.path.join(NGINX_ORIGINAL_ALERT_RULES_PATH, "*.*")):
-            shutil.copy(filename, f"{CONSOLIDATED_ALERT_RULES_PATH}/")
+    def _pull(self, path: str) -> Optional[str]:
+        """Pull file from container (without raising pebble errors).
+
+        Returns:
+            File contents if exists; None otherwise.
+        """
+        try:
+            return cast(str, self._nginx_container.pull(path, encoding="utf-8").read())
+        except (FileNotFoundError, PebbleError):
+            # Drop FileNotFoundError https://github.com/canonical/operator/issues/896
+            return None
+
+    def _push(self, path: str, contents: Any):
+        """Push file to container, creating subdirs as necessary."""
+        self._nginx_container.push(path, contents, make_dirs=True, encoding="utf-8")
+
+    def _push_alert_rules(self, alerts: Dict[str, Any]) -> List[str]:
+        """Pushes alert rules from a rules file to the nginx container.
+
+        Args:
+            alerts: a dictionary of alert rule files, fetched from
+                either a metrics consumer or a remote write provider.
+        """
+        paths = []
+        for topology_identifier, rules_file in alerts.items():
+            filename = f"juju_{topology_identifier}.rules"
+            path = f"{RULES_DIR}/{filename}"
+
+            rules = yaml.safe_dump(rules_file)
+
+            self._push(path, rules)
+            paths.append(path)
+            logger.debug("Updated alert rules file %s", filename)
+
+        return paths
+
+    def _get_mimirtool(self):
+        """Copy the `mimirtool` binary to the workload container."""
+        with open("mimirtool", "rb") as f:
+            self._nginx_container.push("/usr/bin/mimirtool", source=f, permissions=0o744)
+
+    def _set_alerts(self):
+        """Create alert rule files for all Mimir consumers."""
+
+        def sha256(hashable: Any) -> str:
+            """Use instead of the builtin hash() for repeatable values."""
+            if isinstance(hashable, str):
+                hashable = hashable.encode("utf-8")
+            return hashlib.sha256(hashable).hexdigest()
+
+        if not self._nginx_container.can_connect():
+            return
+
+        # Get mimirtool if this is the first execution
+        if not self._pull(ALERTS_HASH_PATH):
+            self._get_mimirtool()
```

Comment:
Seems odd to push mimirtool if no alert hash. The two are technically unrelated.
(mimirtool was already pushed on pebble ready.)
Am I missing something?

Reply:
`self._set_alerts()` is called in the constructor, so I believe it can be called before `_on_pebble_ready`. That `if` is meant to make sure we always have `mimirtool` on the first execution.

Is there a better way to do this?